### ISSUE TITLE
Updating an existing asset

### ISSUE THREAD

**Issue by ashutosh8** [contentful/contentful-management.js#68]
Hi,

I was trying to update an existing asset. The api throws an error "_VersionMismatch_" when I try to update the asset with method "asset.update" followed by "asset.publish".

What I inferred was that between update and publish there is a change of version of the asset. 

Also, the method "space.createAssetWithId" does not take header as an argument along with the body data which could have helped in version management. Since it is a put request, both create and update would have worked.

So is there any method by which asset update can be achieved along with publishing?

**Comment by Khaledgarbaya** [comment]
hi @ashutosh8 ,
The version management is done by the SDK. when you update using `asset.update` and asset you need to call `asset. processForAllLocales ` and then `asset.publish`

**Comment by ashutosh8** [comment]
Thanks @Khaledgarbaya !

I tried doing that as well but still I get the same error. Notably, the issue occurs **only** when I select multiple locales of a particular asset.

**Comment by Khaledgarbaya** [comment]
@ashutosh8 it might be that you need to wait a little bit after processing

```js
var prePublishDelay = 500
           space.creationAsset()
          .delay(prePublishDelay)
          .then((asset) = {
              asset.processForAllLocale()
          })
          .delay(prePublishDelay)
          .then((asset) = > {
             asset.publish()
          })
```

this is the approach I am doing in the space sync tool we have

**Comment by ashutosh8** [comment]
Hi @Khaledgarbaya,

I understand that a bit delay would help it get published. But that delay is not working in the code.

Also, my application makes requests asynchronously, and I have multiple locales for each asset.

Consider a case for updating an asset with multiple assets but not all locales. Then calling processForAllLocales() wont be successful since the API returns that the asset is already processed. Since it started processing the locale which already has "url" property and not "upload" property.

What I think I need to do is now call processForLocale(locale) for only those locales which require to be processed successively.
What do you suggest can be a better way in this case?

**Comment by axe312ger** [comment]
Hey @ashutosh8,

`processForAllLocales` is just a shorthand for calling `processForLocale()` for every locale in the asset. It should work fine for your usecase.

Were you able to fix your problem or do you still need help?

All the best,
Benedikt

**Comment by Khaledgarbaya** [comment]
I am closing this issue since it is not active anymore

**Comment by soliveira-nascent** [comment]
We're having the same issue where we're getting a versionMismatch when trying to publish an asset after calling processForAllLocales. We are processing an asset for two locales (en-US, fr-CA). 

See example below:

```javascript
space.createAsset({ fields: rawAsset })
      .then((asset) => asset.processForAllLocales())
      .then((asset) => asset.publish())
      .then((asset) => ({ space, asset }))
      .catch((error) => Promise.reject(error))
```

The issue is intermittent. On occasion this succeeds and at other times it fails. After looking at the processForAllLocales method, it appears to be returning the first object in the array of processed asset objects. My guess would be that the first object may contain an outdated version and therefore causing the versionMismatch when trying to publish.

The code I'm referring to is in `lib/entities/asset.js` line 198

```javascript
return Promise.all(locales.map((locale) => processForLocale.call(self, locale, options)))
.then((assets) => assets[0])
```

As a workaround we've add a `space.getAsset(asset.sys.id)` before calling `asset.publish()`, but ideally we shouldn't have to make this call

**Comment by ashutosh8** [comment]
Hi @axe312ger , @Khaledgarbaya ,

`processForAllLocales` works fine when we need to process all the locales of an asset but our usecase involved selected locales to be processed. Hence, we used `processForLocale(locale)`.

The only place where `processForAllLocales`  is used is during creating a new asset with multiple locales.
Although the processing works but publishing was an issue with 'VersionMismatch' error, thus we made a get request to have the latest version of the processed asset and then publish it.

Moreover, issue no. **101** deals with the same scenario of 'VersionMismatch' when using `processForAllLocales`. I hope that would help solve the issue.

Best,
Ashutosh
